src/Kestrel.Core/Internal/Http/OutputProducer.cs

```diff
@@ -148,6 +149,47 @@ public void Abort(Exception error)
             return FlushAsync(writableBuffer, cancellationToken);
         }
 
+        public async Task WriteAsync(
+            Stream stream,
+            long length,
+            CancellationToken cancellationToken)
+        {
+
+            var writableBuffer = default(WritableBuffer);
+            var remaining = length;
+            int count;
+            do
+            {
+                lock (_contextLock)
+                {
+                    if (_completed)
+                    {
+                        return;
+                    }
+
+                    writableBuffer = _pipeWriter.Alloc(1);
+                }
+                
+                // Requires "Add Memory-based Stream overloads to coreclr"
+                // https://github.com/dotnet/coreclr/pull/13769
+                // Read the filestream directly into the output buffer
+                var buffer = writableBuffer.Buffer;
+                if (buffer.Length > remaining)
+                {
+                    buffer = buffer.Slice((int)remaining);
+                }
+
+                count = await stream.ReadAsync(buffer, cancellationToken);
+                remaining -= count;
+                if (count > 0)
+                {
+                    writableBuffer.Advance(count);
+
+                    await writableBuffer.FlushAsync(cancellationToken);
+                }
+            } while (count > 0);
```

Comment:
What should happen when we reach EOF before we read `count` (SendFileAsync arg)? Should this throw?

Reply:
Yes, both offset and count must be in range. Otherwise you'd end up sending a different amount of data than the caller expected you to and this would break Content-Length.